include/SH3/camera/camera.hpp

```diff
@@ -0,0 +1,208 @@
+        FIRST_PERSON
+    };
+
+    /**
+     *  Our camera in 3D space. Keeps track of position, intended position, rotation etc.
+     *
+     *  Supports clipping in a 3D scene (that is, it can collide as well as a 3D model).
+     */
+    struct Camera final
+    {
+    public:
+
+        /**
+         *  Constructor
+         *
+         *  @param pos Initial position of the camera.
+         *  @param fov Camera Horizontal field of view.
+         *  @param aspect Aspect ratio of screen.
+         *  @param zNear Near Z value for frustum culling.
+         *  @param zFar Far Z value for frustum culling.
+         */
+        Camera(const glm::vec3& pos, angle<float> fov, float aspect, float zNear, float zFar);
+
+        /**
+         *  Set the physical position of the camera.
+         *
+         *  @param pos Position vector in 3D space we want the camera at.
+         */
+        void SetPosition(const glm::vec3& pos);
+
+        /**
+         *  Look at position @ref look in 3D space.
+         *
+         *  @param look 3D vector representing the position we want the camera to look at.
+         */
+        void LookAt(const glm::vec3& look);
+
+        /**
+         *  Translates the cameras current position by adding to the position vector.
+         *
+         *  @param trans Our translation vector.
+         */
+        void Translate(const glm::vec3& trans);
+
+        /**
+         *  Translates the cameras current position by adding the camera's Front Vector,
+         *  @ref camFront, and then multiplying by a scale (which effectively chooses)
+         *  the direction.
+         *
+         *  @param factor Translation factor.
+         */
+        void Translate(const float factor);
+
+        /**
+         *  Set the mode of the camera.
+         *
+         *  @param @ref MODE we want the camera to be in.
+         */
+        void SetMode(MODE mode);
+
+        /**
+         *  Set the horizontal FOV of the camera's projection matrix.
+         *
+         *  @param fov Field of view we want the camera to be.
+         */
+        void SetFOV(angle<float> fov);
+
+        /**
+         *  Yaw the camera. There is currently no protection against gimbal lock!
+         *
+         *  @param yaw Yaw to add to the camera's current yaw.
+         */
+        void AddYaw(angle<float> yaw);
+
+        /**
+         *  Pitch the camera. There is currently no protection against gimbal lock!
+         *
+         *  @note Pitch is clamped between [-89, 89] degrees!
+         *
+         *  @param pitch Pitch to add to the camera's current pitch.
+         */
+        void AddPitch(angle<float> pitch);
+
+        /**
+         *  Calculate the projection matrix of our camera by using glm::lookat.
+         *
+         *  @return glm::mat4 that is our projection matrix. Sent to the shader as part of the MVP matrix (it is the P part).
+         */
+        glm::mat4 GetProjectionMatrix() const;
+
+        /**
```

Comment:
This comment block will be merged with the next one for Doxygen. Use `#if 0` instead and "hide" the comment block along with the method declaration. The method declaration then also doesn't need to be commented.